File: 2_二分算法/leetcode2563.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
private:
    int lower_bound(vector<int>& nums, int left, int right, int target) {
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] >= target) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }

public:
    long long countFairPairs(vector<int>& nums, int lower, int upper) {
        // 首先对数组进行排序
        sort(nums.begin(), nums.end());
        long long ans = 0;
        int n = nums.size();
        
        for (int i = 0; i < n; i++) {
            int v = nums[i];
            int low_target = lower - v;
            int high_target = upper - v + 1;              
            int l_index = lower_bound(nums, 0, i, low_target);   // 第一个 >= lower-v 的位置
            int r_index = lower_bound(nums, 0, i, high_target);  // 第一个 >= upper-v+1 的位置          
            r_index--;        
            ans += (r_index - l_index + 1); 

        }
        return ans;
    }
};

File: 2_二分算法/leetcode2563.py
from typing import List


class Solution:
    def lower_bound(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] >= target:
                right = mid - 1
            else:
                left = mid + 1
        return left

    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        nums.sort()
        n = len(nums)
        ans = 0
        for i, v in enumerate(nums):
            ## 对于v先统计左边界
            l = self.lower_bound(nums[0:i], lower - v)
            r = self.lower_bound(nums[0:i], upper - v + 1) - 1
            ans += r - l + 1
        return ans


if __name__ == "__main__":
    a = Solution()
    print(a.countFairPairs([0, 1, 7, 4, 4, 5], 3, 6))
    a = Solution()
